clang/lib/AST/PreorderAST.cpp

```diff
@@ -560,6 +564,81 @@ bool PreorderAST::GetDerefOffset(Node *UpperNode, Node *DerefNode,
   return true;
 }
 
+bool PreorderAST::GetExprIntDiff(Node *E1, Node *E2, llvm::APSInt &Offset) {
+  // Get the integer difference between expressions.
+
+  // If E1 and E2 are not comparable, return false.
+  // Else perform E1 - E2, store the integer result in Offset and return true.
+
+  // E1 and E2 are not comparable if their non-integer parts are not equal.
+
+  // Since we have already normalized exprs like "*p" to "*(p + 0)" we require
+  // that the root of the preorder AST is a BinaryOperatorNode.
+  auto *B1 = dyn_cast_or_null<BinaryOperatorNode>(E1);
+  auto *B2 = dyn_cast_or_null<BinaryOperatorNode>(E2);
+
+  if (!B1 || !B2)
+    return false;
+
+  // If the opcodes mismatch we cannot have a valid offset.
+  if (B1->Opc != B2->Opc)
+    return false;
+
+  // Offset should always be of the form (ptr + offset). So we check for
+  // addition.
+  // Note: We have already converted (ptr - offset) to (ptr + -offset). So
+  // it is okay to only check for addition.
+  if (B1->Opc != BO_Add)
+    return false;
+
+  // We have already constant folded the constants. So return false if the
+  // number of children mismatch.
+  if (B1->Children.size() != B2->Children.size())
+      return false;
+
+  // Check if the children are equivalent.
+  for (size_t I = 0; I != B1->Children.size(); ++I) {
+    auto *Child1 = B1->Children[I];
+    auto *Child2 = B2->Children[I];
+
+    if (Child1->Compare(Child2, Lex) == Result::Equal)
+      continue;
```

Comment:
The output parameter `Offset` is not assigned any valid value in the case when `E1` and `E2` are exactly equal, and the loop that iterates through the children nodes "continues" (at line 605) in each iteration.